### PR TITLE
Save bounce statistics [MAILPOET-3765]

### PR DESCRIPTION
**Pull request by pavel-mailpoet**
[MAILPOET-3765]

[MAILPOET-3765]: https://mailpoet.atlassian.net/browse/MAILPOET-3765

### CODE REVIEW

File: lib/Newsletter/Sending/SendingQueuesRepository.php [hunk 1]
```diff
@@ -45,4 +46,36 @@ public function isSubscriberProcessed(SendingQueueEntity $queue, SubscriberEntit
...
+    SubscriberEntity $subscriber,
+    ?\DateTimeInterface $dateTo,
+    ?\DateTimeInterface $dateFrom
+  ): array {
+    $qb = $this->entityManager->createQueryBuilder()
+      ->select('s, n')
+      ->from(SendingQueueEntity::class, 's')
+      ->join('s.task', 't')
+      ->join('t.subscribers', 'tsub')
+      ->join('tsub.subscriber', 'sub')
```
Comment (lysyjan) [inline lib/Newsletter/Sending/SendingQueuesRepository.php:63]: I think that you don't need this join. You can check the subscriber by the previous join with the alias `tsub`.
Something like this `->andWhere(`tsub.subscriber = :subscriber')`

File: lib/Cron/Workers/Bounce.php [hunk 2]
```diff
@@ -118,4 +132,23 @@ public function getNextRunDate() {
...
+
+  private function findPreviousTask(ScheduledTask $task): ?ScheduledTaskEntity {
+    $taskEntity = $this->scheduledTasksRepository->findOneById($task->id);
+    if (!$taskEntity instanceof ScheduledTaskEntity) return null;
+    return $this->scheduledTasksRepository->findPreviousTask($taskEntity);
+  }
+
+  private function saveBouncedStatistics(SubscriberEntity $subscriber, ScheduledTask $task, ?ScheduledTaskEntity $previousTask) {
+    $taskEntity = $this->scheduledTasksRepository->findOneById($task->id);
+    if (!$taskEntity instanceof ScheduledTaskEntity) return null;
```
Comment (lysyjan) [inline lib/Cron/Workers/Bounce.php:152]: Only a detail here. Here is returned `null`, but the rest of the code returns nothing. You can use a return type declaration `void` and remove `null` here.

**Review by lysyjan — COMMENTED** [review]
It looks good. I have only a few small notes.

File: lib/Newsletter/Sending/ScheduledTasksRepository.php [hunk 3]
```diff
@@ -59,6 +59,19 @@ public function findScheduledOrRunningTask(?string $type): ?ScheduledTaskEntity
     return $queryBuilder->getQuery()->getOneOrNullResult();
   }
 
+  public function findPreviousTask(ScheduledTaskEntity $task): ?ScheduledTaskEntity {
+    return $this->doctrineRepository->createQueryBuilder('st')
```
[the remark this answers is not in the document]
Comment (pavel-mailpoet) [reply lib/Newsletter/Sending/ScheduledTasksRepository.php:63]: @lysyjan 

I tested the query on mailpoet.com and it seems fine: https://d.pr/i/4UsawA https://d.pr/i/3ybEJR
Do you think there is a problem? If so, do you have any suggestion how to improve the performance?

[the remark this answers is not in the document]
Comment (lysyjan) [reply lib/Newsletter/Sending/ScheduledTasksRepository.php:63]: The performance on mailpoet.com is significantly better than on my local. So, I'm fine with the current state.
Thanks!

**Review by lysyjan — APPROVED** [review]

**Review by veljkho — APPROVED** [review]

**Pull request merged**